File: 230_Kth Smallest Element in a BST.java
// 时间复杂度：O(n)，其中 n 是二叉树的大小（节点个数）。
// 空间复杂度：O(h)，其中 h 是树高，递归需要 O(h) 的栈空间。最坏情况下树是一条链，h=n，空间复杂度为 O(n)。

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int k;

    private int inorder(TreeNode node) {
        if (node == null) return -1;

        int left = this.inorder(node.left);
        if (left != -1) return left;

        this.k -= 1;
        if (this.k == 0) return node.val;

        return this.inorder(node.right);
    }

    public int kthSmallest(TreeNode root, int k) {
        this.k = k;
        return this.inorder(root);
    }
}

File: 230_Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # In-Order
        k -= 1  # make it 0-index

        def inorder(node: Optional[TreeNode]):
            # 1. end case
            if not node: return -1

            # 2. process left
            left = inorder(node.left)
            if left != -1: return left

            # 3. process cur-node
            nonlocal k
            if k == 0: return node.val
            k -= 1

            # 4. process right
            return inorder(node.right)
        
        return inorder(root)